Userland/Libraries/LibWeb/HTML/Window.cpp

```diff
@@ -1395,4 +1400,166 @@ JS::NonnullGCPtr<CustomElementRegistry> Window::custom_elements()
+            if (!navigable_container.content_navigable())
+                return IterationDecision::Continue;
+            if (objects.navigables.contains_slow(JS::NonnullGCPtr { *navigable_container.content_navigable() })) {
+                container = navigable_container;
+                return IterationDecision::Break;
+            }
+            return IterationDecision::Continue;
+        });
+        // 2. Return container's content navigable's active WindowProxy.
+        VERIFY(container);
+        return container->content_navigable()->active_window_proxy();
+    }
+
+    // 3. Otherwise, if objects has only one element, return that element.
+    if (objects.elements.size() == 1)
+        return objects.elements[0];
+
+    // 4. Otherwise return an HTMLCollection rooted at window's associated Document,
+    //    whose filter matches only named objects of window with the name name. (By definition, these will all be elements.)
+    return DOM::HTMLCollection::create(const_cast<Window&>(*this).associated_document(), DOM::HTMLCollection::Scope::Descendants, [&name](auto& element) -> bool {
+        if ((is<HTMLEmbedElement>(element) || is<HTMLFormElement>(element) || is<HTMLImageElement>(element) || is<HTMLObjectElement>(element))
+            && (element.attribute(AttributeNames::name) == name.view()))
+            return true;
+        return element.attribute(AttributeNames::id) == name.view();
+    });
+}
+
+// https://html.spec.whatwg.org/#dom-window-nameditem-filter
+Window::NamedObjects Window::named_objects(StringView name)
+{
+    // NOTE: Since the Window interface has the [Global] extended attribute, its named properties
+    //       follow the rules for named properties objects rather than legacy platform objects.
+
+    // Named objects of Window object window with the name name, for the purposes of the above algorithm, consist of the following:
+    NamedObjects objects;
+
+    // document-tree child navigables of window's associated Document whose target name is name;
+    auto children = associated_document().document_tree_child_navigables();
+    for (auto& navigable : children) {
+        if (navigable->target_name() == name) {
+            objects.navigables.append(*navigable);
+        }
+    }
+
+    // embed, form, img, or object elements that have a name content attribute whose value is name
+    // and are in a document tree with window's associated Document as their root; and
+    // HTML elements that have an id content attribute whose value is name and are in a document tree with window's associated Document as their root.
+    associated_document().for_each_in_subtree([&objects, &name, this](auto& node) -> IterationDecision {
+        if (is<HTMLElement>(node)) {
+            auto& element = static_cast<HTMLElement&>(node);
+            if (&element.document() != &associated_document())
+                return IterationDecision::Continue;
```

Comment:
Is it guaranteed that iteration of a document with "for each in subtree" will always give me elements with that document as the root of the tree? Or does this iterate into iframes as well?

Reply:
`Node::for_each_in_subtree` will not traverse into iframes.